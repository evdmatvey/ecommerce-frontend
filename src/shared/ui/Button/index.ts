export { default } from './ui/Button';
