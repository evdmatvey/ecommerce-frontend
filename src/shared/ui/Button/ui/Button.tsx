import clsx from 'clsx';
import { type ComponentProps, type ReactNode } from 'react';
import styles from './Button.module.css';
import { buttonTestIds } from './__tests__/Button.spec';

interface ButtonProps extends ComponentProps<'button'> {
  color?:
    | 'primary'
    | 'secondary'
    | 'info'
    | 'success'
    | 'warning'
    | 'danger'
    | 'black'
    | 'gray';
  size?: 'small' | 'medium' | 'large';
  variant?: 'outlined' | 'solid';
  icon?: ReactNode;
  iconPosition?: 'left' | 'right' | 'only';
  children?: ReactNode;
}

const Button = ({
  color = 'primary',
  size = 'medium',
  variant = 'solid',
  iconPosition = 'left',
  icon,
  children,
  ...props
}: ButtonProps) => {
  const classes = clsx(
    props.className,
    styles.root,
    styles[size],
    styles[color],
    {
      [styles.solid]: variant === 'solid',
      [styles.outlined]: variant === 'outlined',
      [styles.iconOnly]: iconPosition === 'only',
      [styles.icon]: icon,
    },
  );
  return (
    <button className={classes} {...props} data-testid={buttonTestIds.BUTTON}>
      {iconPosition === 'left' && (
        <span data-testid={buttonTestIds.ICON_LEFT}>{icon}</span>
      )}
      {iconPosition === 'only' ? (
        <span data-testid={buttonTestIds.ICON_ONLY}>{icon}</span>
      ) : (
        <span data-testid={buttonTestIds.CHILDREN}>{children}</span>
      )}
      {iconPosition === 'right' && (
        <span data-testid={buttonTestIds.ICON_RIGHT}>{icon}</span>
      )}
    </button>
  );
};

export default Button;
