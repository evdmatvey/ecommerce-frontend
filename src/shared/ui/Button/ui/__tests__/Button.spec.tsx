import { render, screen } from '@testing-library/react';
import { describe, expect, it } from 'vitest';
import Button from '../Button';

export const buttonTestIds = {
  BUTTON: 'button',
  CHILDREN: 'children',
  ICON_LEFT: 'icon-left',
  ICON_ONLY: 'icon-only',
  ICON_RIGHT: 'icon-right',
};

const TEST_TEXT = 'Click me';

const Icon = <span>Icon</span>;

describe('Button icons', () => {
  it('Renders the button with the correct text', () => {
    render(<Button>{TEST_TEXT}</Button>);
    const button = screen.getByTestId(buttonTestIds.BUTTON);
    expect(button).toHaveTextContent(TEST_TEXT);
  });

  it('Renders the button with an icon on the left', () => {
    render(
      <Button icon={Icon} iconPosition="left">
        {TEST_TEXT}
      </Button>,
    );
    const icon = screen.getByTestId(buttonTestIds.ICON_LEFT);
    const children = screen.getByTestId(buttonTestIds.CHILDREN);
    expect(icon).toBeInTheDocument();
    expect(children).toBeInTheDocument();
  });

  it('Renders the button with an icon on the left (without position declaration)', () => {
    render(<Button icon={Icon}>{TEST_TEXT}</Button>);
    const icon = screen.getByTestId(buttonTestIds.ICON_LEFT);
    const children = screen.getByTestId(buttonTestIds.CHILDREN);
    expect(icon).toBeInTheDocument();
    expect(children).toBeInTheDocument();
  });

  it('Renders the button with an icon only', () => {
    render(<Button icon={Icon} iconPosition="only" />);
    const icon = screen.getByTestId(buttonTestIds.ICON_ONLY);
    expect(icon).toBeInTheDocument();
    expect(
      screen.queryByTestId(buttonTestIds.CHILDREN),
    ).not.toBeInTheDocument();
  });

  it('Renders the button with an icon on the right', () => {
    render(
      <Button icon={Icon} iconPosition="right">
        {TEST_TEXT}
      </Button>,
    );
    const icon = screen.getByTestId(buttonTestIds.ICON_RIGHT);
    const children = screen.getByTestId(buttonTestIds.CHILDREN);
    expect(icon).toBeInTheDocument();
    expect(children).toBeInTheDocument();
  });
});

describe('Button snapshots', () => {
  it('Button with left icon (without position declaration)', () => {
    const { asFragment } = render(<Button icon={Icon}>{TEST_TEXT}</Button>);
    expect(asFragment()).toMatchSnapshot();
  });

  it('Button with left icon', () => {
    const { asFragment } = render(
      <Button icon={Icon} iconPosition="left">
        {TEST_TEXT}
      </Button>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('Button with right icon', () => {
    const { asFragment } = render(
      <Button icon={Icon} iconPosition="right">
        {TEST_TEXT}
      </Button>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('Button with icon only', () => {
    const { asFragment } = render(<Button icon={Icon} iconPosition="only" />);
    expect(asFragment()).toMatchSnapshot();
  });
});
