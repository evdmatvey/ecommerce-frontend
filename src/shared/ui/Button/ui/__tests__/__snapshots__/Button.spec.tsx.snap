// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button snapshots > Button with icon only 1`] = `
<DocumentFragment>
  <button
    class="_root_bfc85f _medium_bfc85f _primary_bfc85f _solid_bfc85f _iconOnly_bfc85f _icon_bfc85f"
    data-testid="button"
  >
    <span
      data-testid="icon-only"
    >
      <span>
        Icon
      </span>
    </span>
  </button>
</DocumentFragment>
`;

exports[`Button snapshots > Button with left icon (without position declaration) 1`] = `
<DocumentFragment>
  <button
    class="_root_bfc85f _medium_bfc85f _primary_bfc85f _solid_bfc85f _icon_bfc85f"
    data-testid="button"
  >
    <span
      data-testid="icon-left"
    >
      <span>
        Icon
      </span>
    </span>
    Click me
  </button>
</DocumentFragment>
`;

exports[`Button snapshots > Button with left icon 1`] = `
<DocumentFragment>
  <button
    class="_root_bfc85f _medium_bfc85f _primary_bfc85f _solid_bfc85f _icon_bfc85f"
    data-testid="button"
  >
    <span
      data-testid="icon-left"
    >
      <span>
        Icon
      </span>
    </span>
    Click me
  </button>
</DocumentFragment>
`;

exports[`Button snapshots > Button with right icon 1`] = `
<DocumentFragment>
  <button
    class="_root_bfc85f _medium_bfc85f _primary_bfc85f _solid_bfc85f _icon_bfc85f"
    data-testid="button"
  >
    Click me
    <span
      data-testid="icon-right"
    >
      <span>
        Icon
      </span>
    </span>
  </button>
</DocumentFragment>
`;
